File: crates/agent/src/state.rs
//! AgentState — plain data (§11: snapshot/restore, pause-and-resume across
//! refreshes all depend on this being serializable, which async-fn futures
//! could never be — ARCHITECTURE §1c).
//!
//! THIS FILE IS THE VOCABULARY: every field the agent has, with the argument
//! for why it exists beside it. [`opening`] holds the one particular VALUE of
//! that vocabulary a fresh agent starts as, in one piece, because "what is an
//! agent before anything has happened to it" is a question with its own answer
//! and reading the record should not mean reading past forty initialisers.

use std::collections::BTreeMap;

use serde::{Deserialize, Serialize};

use context::{Part, State};
use kernel::PhaseId;

use crate::spec::defaults::{
    default_compact_at, default_keep_recent, default_max_rounds, default_passes,
};
use crate::toolbox::Toolbox;

/// One planned step with its own success criteria — Verify judges against
/// these, not vibes (ADR-010: "the judge reads the spec").
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct PlanStep {
    pub description: String,
    pub success_criteria: String,
}

/// The whole agent between events. Everything `step` may consult is a field
/// here; everything not here does not exist to the agent (I7). Serializable
/// because a state that can be written down is a state a refresh can resume
/// (I11) — today through the stored LOG (`core::log::decisions`).
/// (No `Eq`: `temperature` is an `f64`, which has no total equality.)
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct AgentState {
    pub phase: PhaseId,
    /// This agent's `model:` frontmatter key — a MODEL CATALOGUE key, never a
    /// URL (increment 04). It rides out on `Effect::CallModel` so the adapter
    /// can resolve it; empty means "the catalogue's default entry".
    #[serde(default)] pub model: String,
    /// This agent's `temperature:` frontmatter key, or `None` where the file
    /// named none. Rides out on `Effect::CallModel` (increment 19): the key
    /// parsed, rendered back out and printed on the card for eighteen rounds
    /// while reaching no request body — the `compact_at: lots` failure
    /// (`spec::number`) in a key that looked applied.
    #[serde(default)] pub temperature: Option<f64>,
    /// What is being attempted; `None` = idle, awaiting a task.
    pub task: Option<String>,
    /// The current plan; empty in Work-first flows (RESEARCH phase-cut:
    /// Plan-on-demand, not mandatory).
    pub plan: Vec<PlanStep>,
    /// Which plan step Work is on — the LOOP advances this, never the model
    /// (ADR-010: one step per call).
    pub cursor: usize,
    /// Consecutive-retry guard counter; a looping model terminates
    /// deterministically because these live here, not in prose (ADR-010).
    pub retries: u8,
    pub replans: u8,
    /// Tool results still outstanding from the batch the model just wrote.
    /// The model sees none of them until this reaches zero — that is what
    /// makes one line of calls one observation (Python `core/tools.py`).
    #[serde(default)] pub pending_tools: usize,
    /// How many times this turn has already gone round the tool loop. A
    /// looping model terminates on this counter, never on prose.
    #[serde(default)] pub tool_rounds: u16,
    /// A sentence the person typed while this turn was already running, not
    /// yet answered. It is a FLAG and not a queue: the sentence itself is
    /// already in the history, and this only records that nothing has replied
    /// to it since. Cleared the moment a call is made carrying it.
    #[serde(default)] pub steered: bool,
    /// The person pressed Stop while this turn was running. A FLAG, like
    /// `steered`, and for the same reason: the press is a fact about the turn
    /// and not a queue of work. `stop::boundary` consumes it at the next step
    /// boundary; a new turn clears it.
    #[serde(default)] pub stopping: bool,
    /// The ceiling that counter terminates on, from this agent's `max_rounds:`
    /// frontmatter. It is per-agent because the right number is a property of
    /// the WORK: a summarizer that calls two tools and a coding agent that
    /// edits nine files, builds, reads the errors and edits again cannot share
    /// one constant, and the constant this replaced was four.
    #[serde(default = "default_max_rounds")] pub max_rounds: u16,
    /// What THIS agent may call: its `agent.md` `tools:` list resolved
    /// against the built-ins and its peers (`subagent::toolbox_for`). In
    /// state, not in the phase table, because it is the agent's property and
    /// not the machine's — the phase only decides whether this phase may act.
    #[serde(default)] pub toolbox: Toolbox,
    /// Compact once the window reaches this many entries; 0 never compacts
    /// (Python `Engine.compact_at`, default 75, overridable in frontmatter).
    #[serde(default = "default_compact_at")] pub compact_at: usize,
    /// How many of the newest entries survive a compaction verbatim (Python
    /// `Engine.keep_recent`, default 24).
    #[serde(default = "default_keep_recent")] pub keep_recent: usize,
    /// The reply now in flight is the SUMMARIZER's, not this agent's answer.
    /// A summary is an artifact compaction produces and assembly reads back —
    /// `assemble` is pure and cannot author one (I14, RESEARCH).
    #[serde(default)] pub compacting: bool,
    /// How many times this window has been compacted. The log mirrors the
    /// window, and this counter is what tells the mirror a REWRITE is due
    /// rather than another append.
    #[serde(default)] pub compactions: u32,
    /// THIS TURN'S EVIDENCE (`crate::verify`). Two flags folded left-to-right
    /// over the turn's tool results, and `nudges` counting how many times the
    /// gate has already asked. All three are turn-scoped: cleared where
    /// `pending_tools` and `tool_rounds` are, and again when a turn ends.
    #[serde(default)] pub mutated: bool,
    #[serde(default)] pub green: bool,
    #[serde(default)] pub nudges: u8,
    /// The loop this agent's file declares (`crate::stages`), in order, and
    /// how far this turn has walked it. Empty is the react loop alone — which
    /// is every agent written before the key existed, and the reason nothing
    /// about the old single-stage turn changed.
    #[serde(default)] pub stages: Vec<String>,
    /// The list the agent's FILE declares, which `stages` is reset to at the
    /// start of every turn. The two are separate because the strategy stage
    /// REWRITES `stages` mid-turn: without a copy of the declaration, the
    /// second message of a conversation would inherit the route the first one
    /// chose, and a greeting after a project would still be planning.
    #[serde(default)] pub declared: Vec<String>,
    #[serde(default)] pub stage: usize,
    /// The `passes:` budget, the laps spent, and whether THIS lap changed or
    /// ran anything — the continue condition, mechanical and never the model's
    /// verdict (`crate::passes`). 1/0/false is today's turn exactly.
    #[serde(default = "default_passes")] pub passes: u16,
    #[serde(default)] pub pass: u16,
    #[serde(default)] pub acted: bool,
    /// THE SEPARATE REVIEWER (`crate::critic`, 25). `critic` names the agent
    /// holding `role: critic`, so its answer is recognised as a verdict rather
    /// than by a hardcoded name; `reviewed` is what that verdict said — `None`
    /// where it was never asked, or a write since made it stale.
    #[serde(default)] pub critic: String,
    #[serde(default)] pub reviewed: Option<bool>,
    /// THE STANDING GOAL (26) — what this agent's FILE declares it is for, and
    /// what the harness last observed about it. One field and not five because
    /// the three declared lines and the two observations are one mechanism;
    /// `crate::goal` holds it, and the default is an agent that declared none,
    /// which is every agent written before the key and the reason the continue
    /// condition of all of them is untouched.
    #[serde(default)] pub standing: crate::goal::Standing,
    /// The shared space this agent works in, as of the last time it was read
    /// — its facts and notes go into CONTEXT on every call. `None` means the
    /// agent's file named no space, so it works alone (Python: `space` is an
    /// optional frontmatter key).
    #[serde(default)] pub space: Option<crate::space::Space>,
    /// The faculties this agent's file declared (`crate::faculty`), in order.
    /// The host walks this list before every pass and writes fresh state into
    /// `senses` for each. Empty is an agent that senses nothing, which is
    /// every agent written before the key existed.
    #[serde(default)] pub faculties: Vec<String>,
    /// BLOCK ID -> the parts a host wrote for it, most recently. `space` one
    /// step generalised: the slot where an impure host leaves fresh data for a
    /// pure component (`components::Sensed`) to render. `Vec<Part>` and not
    /// `String` so a screenshot is representable without a second mechanism;
    /// `BTreeMap` and not `HashMap` so two identical agents assemble two
    /// identical papers (I7, I14).
    #[serde(default)] pub senses: BTreeMap<String, Vec<Part>>,
    /// WHAT EACH STAGE IS TOLD (`crate::brief`), loaded from `public/stages/`.
    /// Here and not in the spec because a brief is a property of the STAGE —
    /// whose meaning belongs to the machine — and not of the agent, whose own
    /// voice is already the `Soul` block.
    #[serde(default)] pub briefs: crate::brief::Briefs,
    /// The paper's assembly inputs — gathered section sources, refreshed by
    /// `core` (affordances from the registry, observations from effects)
    /// before each step. Inside AgentState so one snapshot restores the
    /// whole thinking context.
    pub paper: State,
}

mod opening;

File: packages/agent/src/state.js
/**
 * AGENTSTATE — the whole agent between events, as plain data.
 *
 * THIS FILE IS THE VOCABULARY: every field the agent has, with the argument for
 * why it exists beside it. Everything `step` may consult is here; anything not
 * here does not exist to the agent (I7). Data and not a class, because a state
 * that can be written down is a state a refresh can resume (I11) — which is
 * also why no field holds a function, a promise or a port.
 *
 * FOUR FIELDS DID NOT SURVIVE THE PORT, MEASURED ON THE RUST TREE: `plan`,
 * `cursor` and `replans` had zero readers and zero writers outside the
 * constructor, and `retries` had two writers that both set it to 0 and no
 * reader — a counter that could only ever hold zero. They were the residue of
 * the phase machine the stage machine replaced. `phase` went with them: it had
 * no assignment outside the constructor and the table it indexed had one row
 * (see `stages.js`).
 * @module
 */

import { StoreError } from '@harness/kernel'
import { NATIVE } from './calls.js'
import { checkField, shapeOf } from './shape.js'

/** @typedef {import('@harness/kernel').TurnId} TurnId */
/** @typedef {import('@harness/context').ModelCard} ModelCard */
/** @typedef {import('./calls.js').CallStyle} CallStyle */
/** @typedef {import('./round.js').Asked} Asked */
/** @typedef {import('./tools.js').Tool} Tool */
/** @typedef {import('./turn.js').Awaiting} Awaiting */

/** The agent's paper — `packages/context`'s assembly input; A-PAPER owns its shape. @typedef {{sources: unknown[]}} Paper */

/** What an agent's file declares it is FOR. `goal.js` (B18) reads it; the shape lives here because it is state. @typedef {{outcome: string, check: string, doneWhen: string}} Goal */

/**
 * The declaration plus what the harness OBSERVED about it on this lap.
 * `checking` says the next tool result is the harness's own question and not
 * the model's; `met` is that check's exit code, null where this lap has not
 * read it. Both are lap-scoped: evidence about a lap that is over says nothing
 * about the one starting.
 * @typedef {{goal: Goal, checking: boolean, met: boolean | null}} Standing
 */

/** The shared folder an agent works in, as of the last read. `space.js` (B27) mutates it. @typedef {{name: string, facts: Array<[string, string]>, notes: string[]}} Space */

/**
 * @typedef {object} AgentState
 * @property {string} name  how this agent is addressed — the `identity` block's own words, and the name a peer delegates to. On the state because a running agent that cannot say what it is called is one the person cannot address either.
 * @property {string} description  its one-line role, as its file declares it; the second half of `identity`
 * @property {string[]} conversation  the window, oldest first, one already-tagged line per turn — what the `history` block renders. It is the thing `compactAt` counts and `keepRecent` spares, and until it was here the assembled prompt carried NO conversation at all: the paper's sources were the only channel and nothing in this build ever wrote one.
 * @property {string} model  this agent's `model:` key — a CATALOGUE key, never a URL; empty means the catalogue's default
 * @property {number | null} temperature  its `temperature:` key, null where the file named none
 * @property {string | null} task  what is being attempted; null is idle
 * @property {TurnId} turnId  WHICH attempt it is, or '' when idle. Every effect is stamped with it and the reducer drops any fact that names another (I21) — the predecessor had no such name, which is how a result from an abandoned turn billed a fresh model call.
 * @property {Awaiting} awaiting  what this turn has outstanding: the model, the tools, or nothing. Explicit because "nothing is outstanding" and "one tool is outstanding" were both `pending_tools == 0`, and a result arriving against nothing awaited must be an anomaly rather than a new request.
 * @property {Tool[]} toolbox  what THIS agent may call, resolved from its `tools:` list, as DESCRIPTORS (`tools.js`) and not names. The loop needs the schema to refuse a call and the two declared properties to fold what a result proved; a list of names is what made both live somewhere else.
 * @property {CallStyle} calling  how this model's calls arrive: its own API, or read out of the text by the declared fallback. A property of the model card, adopted onto the state — never guessed from what a reply happens to look like.
 * @property {Asked[]} batch  the round the model just wrote, each call with its own id, in the order written. Results are filed against it BY ID; the model sees none of them until every entry is done — that is what makes one round of calls one observation. It replaced a count, because "nothing is outstanding" and "this result answers nothing" were both `pending_tools == 0`.
 * @property {string[]} observations  this round's results, in arrival order. An ARRAY, because the Rust upserted a one-line component by id and three calls on one line therefore produced three overwrites the model saw one of. Replaced when the next round of calls is written.
 * @property {number} toolRounds  how many times this turn has gone round the tool loop; a looping model terminates on this counter, never on prose
 * @property {number} maxRounds  the ceiling that counter terminates on. Per-agent because the right number is a property of the WORK: a summarizer calling two tools and a coding agent editing nine files cannot share one constant, and the constant this replaced was four.
 * @property {boolean} steered  a sentence typed into a running turn, not yet answered. A FLAG and not a queue: the sentence is already in the history, and this only records that nothing has replied to it.
 * @property {boolean} stopping  the person pressed Stop mid-turn. A flag for the same reason; consumed at the next step boundary, cleared by a new turn.
 * @property {number} compactAt  compact once the window reaches this many entries; 0 never compacts
 * @property {number} keepRecent  how many of the newest entries survive a compaction verbatim
 * @property {boolean} compacting  the reply in flight is the SUMMARIZER's, not this agent's answer — assembly is pure and cannot author a summary (I14)
 * @property {number} compactions  how many times this window has been compacted; the log mirrors the window, and this is what tells the mirror a REWRITE is due rather than another append
 * @property {boolean} mutated  this turn wrote something (turn-scoped evidence, folded left to right over the tool results)
 * @property {boolean} green  something ran after that write and said so
 * @property {number} nudges  how many times the verify gate has already asked
 * @property {string[]} stages  the loop this turn is walking, in order. Empty is the react loop alone — every agent written before the key existed.
 * @property {string[]} declared  the list the agent's FILE declares, which `stages` is reset to each turn. Separate because the strategy stage REWRITES `stages` mid-turn: without the copy, a greeting after a project would still be planning.
 * @property {number} stage  how far this turn has walked that list
 * @property {number} passes  how many times one turn may walk the list
 * @property {number} pass  laps spent
 * @property {boolean} acted  whether THIS lap changed or ran anything — the continue condition, mechanical and never the model's verdict
 * @property {string} critic  the agent holding `role: critic`, so its answer is recognised as a verdict rather than by a hardcoded name
 * @property {boolean | null} reviewed  what that verdict said; null where it was never asked, or a write since made it stale
 * @property {Standing} standing  the standing goal and this lap's observation of it
 * @property {Space | null} space  null means the file named no space, so the agent works alone
 * @property {string[]} faculties  the faculties the file declared, in order; the host refreshes `senses` for each before every pass
 * @property {Record<string, unknown[]>} senses  BLOCK ID -> the parts a host wrote for it, most recently. The slot where an impure host leaves fresh data for a pure component to render. Parts and not strings so a screenshot needs no second mechanism.
 * @property {Record<string, string>} briefs  what each stage is TOLD, loaded from `public/stages/`. Here and not in the spec because a brief is a property of the STAGE, whose meaning belongs to the machine.
 * @property {Paper} paper  the assembly inputs, refreshed before each step. Inside the state so one snapshot restores the whole thinking context.
 * @property {string} prompt  this agent's file's own markdown body — the words the soul block is written in. On the state and not fetched per call: the paper is derived every turn and the agent's own words are an input to it, not a lookup.
 * @property {ModelCard | null} card  what this agent's `model:` key resolved to in the catalogue — the window every budget is derived from. Null is a file naming a model this build has no entry for, and the turn ends saying so rather than assembling against a number somebody invented.
 * @property {number} attempts  how many times THIS turn's model call has failed and been asked again. Turn-scoped: a provider that failed twice yesterday says nothing about the call going out now.
 * @property {string} lastEmpty  the `model|finish` pair of the previous zero-output completion, '' when the last reply carried something. Two identical ones in a row stop the retry — a model answering deterministically will answer the same nothing again.
 */

/** Sixty-four, not four: four rounds cannot finish real work — read, build, read the errors, edit, build is already five — and the ceiling exists to stop a model LOOPING, not to stop an agent working. */
export const DEFAULT_MAX_ROUNDS = 64
export const DEFAULT_COMPACT_AT = 75
export const DEFAULT_KEEP_RECENT = 24
/** ONE, and one is not a placeholder: one pass is byte-for-byte the turn this build has always taken. A file that wants the loop asks for it. */
export const DEFAULT_PASSES = 1

/**
 * A fresh idle agent. Nothing else may build an AgentState from parts, so
 * "what an absent host has left behind" is written once: no spec adopted, so
 * no tools — the honest default is that nothing is attached an agent did not
 * ask for. No briefs either: seeding them here would be the compiled-in
 * fallback that makes a missing `public/stages/` file invisible.
 *
 * NO ARGUMENTS. A seeded paper was a parameter every caller passed the same
 * value for; the code that would have a real one to give does not exist yet,
 * and when it does it adopts one through `ask.js`, which
 * derives the whole paper per call, rather than through an optional argument
 * here.
 * @returns {AgentState}
 */
export function newAgentState() {
  return {
    name: '', description: '', conversation: [],
    model: '', temperature: null, task: null, turnId: '', awaiting: null, toolbox: [],
    calling: NATIVE, batch: [], observations: [], toolRounds: 0, maxRounds: DEFAULT_MAX_ROUNDS,
    steered: false, stopping: false,
    compactAt: DEFAULT_COMPACT_AT, keepRecent: DEFAULT_KEEP_RECENT,
    compacting: false, compactions: 0,
    mutated: false, green: false, nudges: 0,
    stages: [], declared: [], stage: 0,
    passes: DEFAULT_PASSES, pass: 0, acted: false,
    critic: '', reviewed: null,
    standing: { goal: { outcome: '', check: '', doneWhen: '' }, checking: false, met: null },
    space: null, faculties: [], senses: {}, briefs: {}, paper: { sources: [] },
    prompt: '', card: null, attempts: 0, lastEmpty: '',
  }
}

/**
 * The state as bytes, canonically. Keys are sorted at every depth so two
 * identical agents write identical records — the Rust used a `BTreeMap` for
 * `senses` and `briefs` for exactly this reason, and a JS object's insertion
 * order would otherwise make byte-identity depend on the order a host happened
 * to write its blocks in.
 * @param {AgentState} state @returns {string}
 */
export function serializeAgentState(state) {
  return JSON.stringify(state, (_key, value) => {
    if (value === null || typeof value !== 'object' || Array.isArray(value)) return value
    const entries = Object.entries(value).sort(([a], [b]) => (a < b ? -1 : 1))
    return Object.fromEntries(entries)
  })
}

/**
 * One stored state, read back. Absent keys take the fresh default, which is how
 * a record written before a field existed still loads.
 *
 * Everything else REFUSES, by name (I18). A key this build has no field for
 * means a record from a newer build, and a key of the wrong shape means a
 * record this reader cannot understand; guessing at either would put an agent
 * on the page in a state no code was written for. `shape.js` decides the second
 * half, and it checks THROUGH the compound fields.
 *
 * `Object.hasOwn` and not `in`: `in` walks the prototype chain, so a record
 * carrying `__proto__` or `toString` would answer "this build has that field",
 * be waved past, and then be silently dropped by the spread below — which is
 * the one thing I18 says a reader may never do.
 * @param {string} text @returns {AgentState}
 */
export function restoreAgentState(text) {
  const raw = readRecord(text)
  const fresh = newAgentState()
  for (const key of Object.keys(raw)) {
    if (Object.hasOwn(fresh, key)) continue
    throw new StoreError('corrupt', `This agent state was written by a newer build: it carries "${key}", which this one has no field for.`, { key })
  }
  for (const [key, value] of Object.entries(raw)) {
    const bad = checkField(key, value, /** @type {Record<string, unknown>} */ (fresh)[key])
    if (!bad) continue
    throw new StoreError('corrupt', `This agent state holds "${bad.key}" as ${bad.found}, and this build reads it as ${bad.want}.`, { key: bad.key })
  }
  // Every key was proved present in `fresh` and every shape was proved above;
  // tsc cannot follow a proof that walks a table, so the cast carries it.
  return /** @type {AgentState} */ ({ ...fresh, ...raw })
}

/** @param {string} text @returns {Record<string, unknown>} */
function readRecord(text) {
  /** @type {unknown} */
  let value
  try {
    value = JSON.parse(text)
  } catch (cause) {
    throw new StoreError('corrupt', 'This agent state is not JSON, so nothing in it can be read.', { cause })
  }
  if (typeof value !== 'object' || value === null || Array.isArray(value)) {
    throw new StoreError('corrupt', `This agent state is ${shapeOf(value)} where a record was expected.`)
  }
  return /** @type {Record<string, unknown>} */ (value)
}
